pkg/metrics/daemon.go

```diff
@@ -0,0 +1,142 @@
+	if idCfg.Init {
+		log.Infof("Metrics exporter is disabled for init mode: address[%s]", idCfg.MetricsServerAddr)
+		return ms, nil
+	}
+	if idCfg.MetricsServerAddr == "" {
+		log.Infof("Metrics exporter is disabled with empty options: address[%s]", idCfg.MetricsServerAddr)
+		return ms, nil
+	}
+
+	// https://github.com/enix/x509-certificate-exporter
+	// https://github.com/enix/x509-certificate-exporter/blob/main/cmd/x509-certificate-exporter/main.go
+	// https://github.com/enix/x509-certificate-exporter/blob/beb88b34b490add4015c8b380d975eb9cb340d44/internal/exporter.go#L26
+	exporter := internal.Exporter{
+		ListenAddress: idCfg.MetricsServerAddr,
+		SystemdSocket: false,
+		ConfigFile:    "",
+		Files: []string{
+			idCfg.CertFile,
+			idCfg.CaCertFile,
+		},
+		Directories:           []string{},
+		YAMLs:                 []string{},
+		TrimPathComponents:    0,
+		MaxCacheDuration:      time.Duration(0),
+		ExposeRelativeMetrics: true,
+		ExposeErrorMetrics:    true,
+		KubeSecretTypes: []string{
+			"kubernetes.io/tls:tls.crt",
+		},
+		KubeIncludeNamespaces: []string{},
+		KubeExcludeNamespaces: []string{},
+		KubeIncludeLabels:     []string{},
+		KubeExcludeLabels:     []string{},
+	}
+
+	if len(idCfg.TargetDomainRoles) != 0 && idCfg.RoleCertDir != "" {
+		for _, dr := range idCfg.TargetDomainRoles {
+			fileName := dr.Domain + idCfg.RoleCertFilenameDelimiter + dr.Role + ".cert.pem"
+			exporter.Files = append(exporter.Files, strings.TrimSuffix(idCfg.RoleCertDir, "/")+"/"+fileName)
+		}
+	}
+
+	ms.exporter = &exporter
+	return ms, nil
+}
+
+// Start starts the metrics exporter
+func (ms *metricsService) Start(ctx context.Context) error {
+	if ctx.Err() != nil {
+		log.Info("Skipped metrics exporter start")
+		return nil
+	}
+
+	if ms.exporter != nil {
+		log.Infof("Starting metrics exporter server[%s]", ms.idCfg.MetricsServerAddr)
+		ms.shutdownWg.Add(1)
+		go func() {
+			defer ms.shutdownWg.Done()
+			if err := ms.exporter.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				log.Fatalf("Failed to start metrics exporter server: %s", err.Error())
+			}
+			log.Info("Stopped metrics exporter server")
+		}()
+
+		if err := daemon.WaitForServerReady(ms.exporter.ListenAddress, false); err != nil {
+			log.Errorf("Failed to confirm metrics exporter server ready: %s", err.Error())
+			return err
+		}
+		ms.exporterRunning = true
+	}
+
+	return nil
+}
+
+func (ms *metricsService) Shutdown() {
+	log.Info("Initiating shutdown of metrics exporter daemon ...")
+	close(ms.shutdownChan)
+
+	if ms.exporter != nil && ms.exporterRunning {
+		err := ms.exporter.Shutdown() // context.Background() is used, no timeout
+		// P.S. Make sure to use the httpChecker to ensure ListenAndServe() is finished before Shutdown() is called. If ListenAndServe() does not finish creating the server object before Shutdown() is called, the internal server field will be nil and Shutdown() be a no-op. ListenAndServe() will block and cause deadlock.
```

Comment:
@WindzCUHK Did you write this message by yourself or just brought it from some where else?
If the message is just brought it from some where else, I suppose we should at least leave the link to the original message.

Reply:
fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/100/commits/ecabacf69caee7f455effe795600728ea71fb361